src/toast/ops/demodulation.py

```diff
@@ -0,0 +1,701 @@
+                demod_detset = []
+                for prefix in self.prefixes:
+                    demod_detset.append(f"{prefix}_{det}")
+                demod_detsets.append(demod_detset)
+        else:
+            for detset in detsets:
+                demod_detset = []
+                for det in detset:
+                    for prefix in self.prefixes:
+                        demod_detset.append(f"{prefix}_{det}")
+                demod_detsets.append(demod_detset)
+        return demod_detsets
+
+    @function_timer
+    def _demodulate_sample_sets(self, obs):
+        sample_sets = obs.all_sample_sets
+        if sample_sets is None:
+            return None
+        demod_sample_sets = []
+        offset = 0
+        for sample_set in sample_sets:
+            demod_sample_set = []
+            for chunksize in sample_set:
+                first_sample = offset
+                last_sample = offset + chunksize
+                demod_first_sample = int(np.ceil(first_sample / self.nskip))
+                demod_last_sample = int(np.ceil(last_sample / self.nskip))
+                demod_chunksize = demod_last_sample - demod_first_sample
+                demod_sample_set.append(demod_chunksize)
+                offset += chunksize
+            demod_sample_sets.append(demod_sample_set)
+        return demod_sample_sets
+
+    @function_timer
+    def _demodulate_intervals(self, obs, demod_obs):
+        if self.nskip == 1:
+            demod_obs.intervals = obs.intervals
+            return
+        times = demod_obs.shared[self.times]
+        for name, ivals in obs.intervals.items():
+            timespans = [[ival.start, ival.stop] for ival in ivals]
+            demod_obs.intervals[name] = IntervalList(times, timespans=timespans)
+        return
+
+    @function_timer
+    def _demodulate_flag(self, flags, wkernel, offset):
+        """Collapse flags inside the filter window and downsample"""
+        """
+        # FIXME: this is horribly inefficient but optimization may require
+        # FIXME: a compiled kernel
+        n = flags.size
+        new_flags = []
+        width = wkernel // 2 + 1
+        for i in range(0, n, self.nskip):
+            ind = slice(max(0, i - width), min(n, i + width + 1))
+            buf = flags[ind]
+            flag = buf[0]
+            for flag2 in buf[1:]:
+                flag |= flag2
+            new_flags.append(flag)
+        new_flags = np.array(new_flags)
+        """
+        # FIXME: for now, just downsample the flags.  Real data will require
+        # FIXME:    measuring the total flag within the filter window
+        flags = flags.copy()
+        # flag invalid samples in both ends
+        flags[: wkernel // 2] |= self.demod_flag_mask
+        flags[-(wkernel // 2) :] |= self.demod_flag_mask
+        new_flags = flags[offset % self.nskip :: self.nskip]
+        return new_flags
+
+    @function_timer
+    def _demodulate_common_flags(self, tod, wkernel, offset):
```

Comment:
This looks like some stale toast2 code...

Reply:
Correct! Removed.